vm/actor/src/builtin/miner/mod.rs

```diff
@@ -248,163 +361,253 @@ impl Actor {
-            st.remove_faults(rt.store(), &declared_recoveries)
+            // Note: We could delay this charge until end of deadline, but that would require more accounting state.
+            let total_penalty_target = undeclared_penalty_target + declared_penalty_target;
+            let unlocked_balance = state.get_unlocked_balance(&rt.current_balance()?);
+            let (vesting_penalty_total, balance_penalty_total) = state
+                .penalize_funds_in_priority_order(
+                    rt.store(),
+                    current_epoch,
+                    &total_penalty_target,
+                    &unlocked_balance,
+                )
                 .map_err(|e| {
-                    ActorError::new(
+                    ActorError::downcast(
+                        e,
                         ExitCode::ErrIllegalState,
-                        format!("failed to remove recoveries from faults: {}", e),
+                        format!(
+                            "failed to unlock penalty for {:?}",
+                            undeclared_penalty_power
+                        ),
                     )
                 })?;
+            penalty_total = &vesting_penalty_total + balance_penalty_total;
+            pledge_delta -= vesting_penalty_total;
 
-            st.remove_recoveries(&declared_recoveries).map_err(|e| {
-                ActorError::new(
-                    ExitCode::ErrIllegalState,
-                    format!("failed to remove recoveries: {}", e),
-                )
-            })?;
+            let deadline_idx = params.deadline;
+            deadlines
+                .update_deadline(rt.store(), params.deadline, &deadline)
+                .map_err(|e| {
+                    actor_error!(
+                        ErrIllegalState,
+                        "failed to update deadline {}: {}",
+                        deadline_idx,
+                        e
+                    )
+                })?;
 
-            // Load info for recovered sectors for recovery of power outside this state transaction.
-            if declared_recoveries.is_empty() {
-                Ok(st.info.sector_size)
-            } else {
-                let mut sectors_by_number: HashMap<SectorNumber, SectorOnChainInfo> =
-                    HashMap::new();
-                for sec in sector_infos {
-                    sectors_by_number.insert(sec.info.sector_number, sec);
-                }
-                declared_recoveries.iter().for_each(|i| {
-                    let key = i as u64;
-                    let s = sectors_by_number.get(&key).cloned().unwrap();
-                    recovered_sectors.push(s);
-                });
-                Ok(st.info.sector_size)
-            }
+            state
+                .save_deadlines(rt.store(), deadlines)
+                .map_err(|e| actor_error!(ErrIllegalState, "failed to save deadlines: {}", e))?;
```

Comment:
these two errors should be downcasted as well (requires previously mentioned change)